### PR TITLE
Metadata REST API closes #787

### LINKED ISSUE #787: Implement Metadata REST API against Metadata Service API

### PR DESCRIPTION
**Pull request by Horneth**

### CODE REVIEW

File: engine/src/main/scala/cromwell/services/ServiceRegistry.scala [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package cromwell.services
+
+import com.typesafe.config.ConfigFactory
+import cromwell.server.WorkflowManagerSystem
+
+object ServiceRegistry extends WorkflowManagerSystem {
+  val serviceRegistryActor = actorSystem.actorOf(ServiceRegistryActor.props(ConfigFactory.load()))
```
Comment (geoffjentry) [inline engine/src/main/scala/cromwell/services/ServiceRegistry.scala]: `ServiceRegistryActor`

File: engine/src/main/scala/cromwell/webservice/MetadataBuilderActor.scala [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+  private case class MetadataForAttempt(attempt: Int, metadata: IndexedJsonValue)
+  /** There's one IndexedJsonValue per attempt, hence the list. */
+  private case class MetadataForIndex(index: Int, metadata: List[IndexedJsonValue])
+
+  implicit def dateTimeOrdering: Ordering[DateTime] = scala.Ordering.fromLessThan(_ isBefore _)
+
+  /** Sort events by timestamp, transform them into IndexedJsonValues, and merge them together. */
+  private def eventsToIndexedJson(events: Seq[MetadataEvent]): IndexedJsonValue = {
+    events match {
+      case empty if empty.isEmpty => IndexedJsonObject.empty
```
Comment (geoffjentry) [inline engine/src/main/scala/cromwell/webservice/MetadataBuilderActor.scala]: not saying to change this, just that the lhs made me laugh

File: engine/src/test/scala/cromwell/webservice/MetadataBuilderActorSpec.scala [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+
+
+  it should "build lexigraphically sorted JSON list structure from dotted key syntax" in {
+    val eventBuilderList = List(
+      ("l1[2]", "l12", DateTime.now),
+      ("l1[k]", "l1k", DateTime.now),
+      ("l1[3]", "l13", DateTime.now),
+      ("l1[i]", "l1i", DateTime.now),
+      ("l1[10]", "l110", DateTime.now),
+      ("l1[j]", "l1j", DateTime.now)
```
Comment (cjllanwarne) [inline engine/src/test/scala/cromwell/webservice/MetadataBuilderActorSpec.scala]: It'd be great to have an in-list update test here too - e.g.
```
("x[1]", "a", _)
("x[2]", "a", _)
("x[3]", "a", _)

("x[2]", "b", _)
("x[3]", "b", _)

("x[3]", "c", _)
```

should become:
```
{
  "x": [ "a", "b", "c" ]
}
```

File: engine/src/main/scala/cromwell/services/ServiceRegistry.scala (hunk 1 above)
Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/services/ServiceRegistry.scala]: Also give the actor a name

File: engine/src/main/scala/cromwell/webservice/CromwellApiService.scala [hunk 4]
```diff
@@ -39,7 +41,7 @@ class CromwellApiServiceActor(val workflowManager: ActorRef, config: Config)
   def receive = runRoute(possibleRoutes)
 }
 
-trait CromwellApiService extends HttpService with PerRequestCreator {
+trait CromwellApiService extends HttpService with PerRequestCreator with ServiceRegistry {
```
Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/webservice/CromwellApiService.scala]: Wait... is this right? I don't think the REST API should also be the service registry!?

**Comment by cjllanwarne** [comment]
Otherwise, great! 👍

File: engine/src/main/scala/cromwell/webservice/CromwellApiService.scala (hunk 4 above)
Comment (Horneth) [inline engine/src/main/scala/cromwell/webservice/CromwellApiService.scala]: `ServiceRegistry` is a trait holding a reference to the singleton actor ref for the service registry. Maybe the name is confusing :/

File: engine/src/test/scala/cromwell/webservice/MetadataBuilderActorSpec.scala (hunk 3 above)
Comment (Horneth) [inline engine/src/test/scala/cromwell/webservice/MetadataBuilderActorSpec.scala]: ok

File: engine/src/main/scala/cromwell/webservice/CromwellApiService.scala (hunk 4 above)
Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/webservice/CromwellApiService.scala]: oh! `ServiceRegistryClient`? I was certainly in the "confused" group!

File: engine/src/main/scala/cromwell/webservice/CromwellApiServiceClient.scala [hunk 5]
```diff
@@ -23,14 +25,14 @@ trait SwaggerService extends SwaggerUiResourceHttpService {
   override def swaggerUiVersion = "2.1.1"
 }
 
-object CromwellApiServiceActor {
+object CromwellApiServiceActorClient {
   def props(workflowManagerActorRef: ActorRef, config: Config): Props = {
-    Props(classOf[CromwellApiServiceActor], workflowManagerActorRef, config)
+    Props(classOf[CromwellApiServiceActorClient], workflowManagerActorRef, config)
```
Comment (gauravs90) [inline engine/src/main/scala/cromwell/webservice/CromwellApiServiceClient.scala]: I believe @mcovarr suggested to not using reflection style Props in this way.

Comment (mcovarr) [inline engine/src/main/scala/cromwell/webservice/CromwellApiServiceClient.scala]: Yes please use the constructor unless there's a compelling reason not to.  If the constructor is refactored this code won't be updated automatically.

File: engine/src/main/scala/cromwell/webservice/MetadataBuilderActor.scala [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+object MetadataBuilderActor {
+  sealed trait MetadataBuilderActorState
+  case object Idle extends MetadataBuilderActorState
+  case object WaitingForMetadata extends MetadataBuilderActorState
+
+  def props(serviceRegistryActor: ActorRef) = {
+    Props(new MetadataBuilderActor(serviceRegistryActor))
+  }
+
+  private val KeySeparator = '.'
```
Comment (gauravs90) [inline engine/src/main/scala/cromwell/webservice/MetadataBuilderActor.scala]: If this is what I think this is, then I've changed it to `:`

Comment (Horneth) [inline engine/src/main/scala/cromwell/webservice/MetadataBuilderActor.scala]: Yes it's what you think :) I'll change it too then

File: engine/src/main/scala/cromwell/webservice/MetadataBuilderActor.scala [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+  private val KeySeparator = ':'
+  private val ListNameCaptureID = "listName"
+  private val ListIndexCaptureID = "listIndex"
+  private val ListMatcher = "^(.+)\\[([a-zA-Z0-9]+)\\]$".r(ListNameCaptureID, ListIndexCaptureID)
+  private val AttemptKey = "attempt"
+  private val ShardKey = "shardIndex"
+
+  // Define how 2 IndexedJsonValues can be summed
+  private implicit val IndexedJsonSemigroup: Semigroup[IndexedJsonValue] = {
+    new Semigroup[IndexedJsonValue] {
```
Comment (gauravs90) [inline engine/src/main/scala/cromwell/webservice/MetadataBuilderActor.scala]: I am so intrigued by this 😃

**Comment by gauravs90** [comment]
👍

**Pull request merged**